source/v1/I2CDetail.cpp

```diff
@@ -0,0 +1,293 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2016 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mbed-drivers/platform.h"
+
+#if DEVICE_I2C && DEVICE_I2C_ASYNCH
+
+#include "mbed-drivers/v1/I2C.hpp"
+#include "core-util/CriticalSectionLock.h"
+#include "core-util/atomic_ops.h"
+#include "core-util/assert.h"
+#include "minar/minar.h"
+
+namespace mbed_drivers {
+using namespace mbed;
+namespace v1 {
+namespace detail {
+
+I2CError I2CResourceManager::post_transaction(I2CTransaction *t) {
+    CORE_UTIL_ASSERT(t != nullptr);
+    if (!t) {
+        return I2CError::NullTransaction;
+    }
+    I2CError rc = validate_transaction(t);
+    if (rc != I2CError::None) {
+        return rc;
+    }
+
+    // This can't be lock free because of the need to call append() on TransactionQueue.
+    mbed::util::CriticalSectionLock lock;
+    I2CTransaction * tx = TransactionQueue;
+
+    if (tx) {
+        tx->append(t);
+    } else {
+        TransactionQueue = t;
+        return start_transaction();
+    }
+    return I2CError::None;
+}
+
+void I2CResourceManager::process_event(uint32_t event) {
+    I2CTransaction * t;
+    CORE_UTIL_ASSERT(TransactionQueue != nullptr);
+    if (!TransactionQueue) {
+        return;
+    }
+    // Get the current item in the transaction queue
+    t = TransactionQueue;
+    if (!t) {
+        return;
+    }
+    // If the event is 0, no further action is required
+    if (!event) {
+        return;
+    }
+    // Fire the irqcallback for the segment
+    t->call_irq_cb(event);
+    // This can't be done with atomics: there are too many side-effects.
+    mbed::util::CriticalSectionLock lock;
```

Comment:
wouldn't be more clear to create a new block for lock? If we add more code in the future, might not be easy to spot the intention that this lock applies until the line 93?